examples/webxr_vr_layers.html

```diff
@@ -185,49 +323,125 @@
-
-						session.updateRenderState( { layers: [ equirectLayer, quadLayer, session.renderState.layers[ 0 ] ] } );
+						const equirectLayer = mediaBinding.createEquirectLayer(
+							video,
+							{
+								space: refSpace,
+								layout: "stereo-left-right",
+								// Rotate by 45 deg to avoid stereo conflict with the 3D geometry.
+								transform: new XRRigidTransform(
+									{},
+									{ x: 0, y: .28, z: 0, w: .96 } )
+							}
+						);
+
+						errorMesh.visible = false;
+						session.updateRenderState( { layers: [ equirectLayer, quadLayerPlain, quadLayerMips, guiLayer, session.renderState.layers[ 0 ] ] } );
 						video.play();
 
 					} );
 
 				}
 
-				// Copy image to canvas as required.
+				// Copy image to layers as required.
 				// needsRedraw is set on creation or if the underlying GL resources of a layer are lost.
-				if ( quadLayer && quadLayer.needsRedraw ) {
+				if ( quadLayerPlain && quadLayerPlain.needsRedraw ) {
 
 					const glBinding = new XRWebGLBinding( session, gl );
-					const glayer = glBinding.getSubImage( quadLayer, frame );
+					const glayer = glBinding.getSubImage( quadLayerPlain, frame );
 					renderer.state.bindTexture( gl.TEXTURE_2D, glayer.colorTexture );
 					gl.pixelStorei( gl.UNPACK_FLIP_Y_WEBGL, true );
-					gl.texSubImage2D( gl.TEXTURE_2D, 0, 0, 0, gl.RGBA, gl.UNSIGNED_BYTE, snellenTexture.image );
+					gl.texSubImage2D( gl.TEXTURE_2D, 0,
+									( snellenConfig.textureSizePx - snellenConfig.widthPx ) / 2,
+									( snellenConfig.textureSizePx - snellenConfig.heightPx ) / 2,
+									snellenConfig.widthPx, snellenConfig.heightPx,
+									gl.RGBA, gl.UNSIGNED_BYTE, snellenTexture.image );
+
+				}
+
+				// Same as above but also gl.generateMipmap.
+				if ( quadLayerMips && quadLayerMips.needsRedraw ) {
+
+					const glBinding = new XRWebGLBinding( session, gl );
+					const glayer = glBinding.getSubImage( quadLayerMips, frame );
+					renderer.state.bindTexture( gl.TEXTURE_2D, glayer.colorTexture );
+					gl.pixelStorei( gl.UNPACK_FLIP_Y_WEBGL, true );
+					gl.texSubImage2D( gl.TEXTURE_2D, 0,
+									( snellenConfig.textureSizePx - snellenConfig.widthPx ) / 2,
+									( snellenConfig.textureSizePx - snellenConfig.heightPx ) / 2,
+									snellenConfig.widthPx, snellenConfig.heightPx,
+									gl.RGBA, gl.UNSIGNED_BYTE, snellenTexture.image );
+					gl.generateMipmap( gl.TEXTURE_2D );
+
+				}
+
+				// Same as above, but guiLayer.needsUpdate is set when the user interacts with the GUI.
+				if ( guiLayer && ( guiLayer.needsRedraw || guiLayer.needsUpdate ) ) {
+
+					const glBinding = new XRWebGLBinding( session, gl );
+					const glayer = glBinding.getSubImage( guiLayer, frame );
+					renderer.state.bindTexture( gl.TEXTURE_2D, glayer.colorTexture );
+					gl.pixelStorei( gl.UNPACK_FLIP_Y_WEBGL, true );
+					const canvas = guiMesh.material.map.image;
+					gl.texSubImage2D( gl.TEXTURE_2D, 0, 0, 0, canvas.width, canvas.height, gl.RGBA, gl.UNSIGNED_BYTE, canvas );
```

Comment:
Sorry but I think all of this overcomplicates the example. We should try to keep it as simple as possible.

Reply:
Would it be clearer if I pulled the boilerplate code into a utility library at examples/jsm/webxr/layers.js? There is no way to demonstrate text rendering to layers without the `gl.texSubImage2D` call. And one of the major use cases of layers is for high quality text and 2D UI rendering.

Reply:
> Would it be clearer if I pulled the boilerplate code into a utility library at examples/jsm/webxr/layers.js?

That sounds like a step in the right direction.